### PR TITLE
Feature/linkta 82 initialize generative model for gemini

### PR DESCRIPTION
**Pull request by dcpavel**
## Description

Create the GeminiModel and the aiFactory utility in order to connect to Gemini and receive a response.

You can send a POST request to /gen-ai/query with a JSON in the format:
`{
  "prompt": "YOUR QUESTION HERE"
}`

and receive a response. We will tackle integrating the user prompt in LINKTA-91.

## Type of change

- [x] ✨  New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Only manual testing for now.

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I have reviewed the [**Security Best Practices Document**](SECURITY_BEST_PRACTICES.md).
- [x] I have followed the [**Naming Conventions Guide**](NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [x] I have ensured that my pull request title is descriptive.

### CODE REVIEW

**Comment by dcpavel** [comment]
I notice that there is branch pollution by pulling in github actions linting that I didn't realize. Closing for now, I will reopen once I resolve this.

**Comment by dcpavel** [comment]
Reopening PR with code up to date.

**Comment by dcpavel** [comment]
Since a few of my PRs were breaking let me clarify how I've tested this:

1. I ran `npm run lint` to catch any linting errors
2. I ran `npm run dev` and the server booted up
3. I sent a POST request to the server with { "prompt": "please tell me a story" } and received this response:
{
    "message": "AI response",
    "response": "In the tranquil town of Willow Creek, where secrets whispered through the rustling leaves, there lived an enigmatic young woman named Anya. With piercing blue eyes that held a haunted depth and a mind as sharp as a razor, she had always been an outsider, observing the world from the sidelines.\n\nOne fateful evening, as a violent storm raged outside, a group of friends dared to venture into the forbidden Everwood Forest. Among them was Anya, drawn by a morbid curiosity that overshadowed her apprehension. As they delved deeper into the shadowy undergrowth, the wind howled like a banshee, tearing through the trees.\n\nSuddenly, a deafening crash split the air, and Anya's heart skipped a beat. They had stumbled upon the wreckage of an old manor, its crumbling facade draped in ivy. Curiosity got the better of them, and they cautiously approached the dilapidated building.\n\nInside, time seemed to have stood still. Dust danced in the dim light, revealing cobwebbed furniture and faded portraits that stared down at them with hollow eyes. As they explored the upper rooms, Anya noticed a peculiar room hidden behind a secret panel.\n\nWith trembling hands, she pushed the door open and gasped. It was the library, lined with shelves filled with ancient books. In the center stood a massive desk, upon which lay a leather-bound journal. Its pages were yellowed with age, and an unsettling aura seemed to emanate from its covers.\n\nUnable to resist the temptation, Anya reached out and opened the journal. The handwriting was spidery and faded, but the words it contained sent shivers down her spine. It was the diary of the manor's former mistress, a woman named Eleanor Hawthorne.\n\nEleanor's words revealed a tale of love, betrayal, and a dark secret that had haunted her family for generations. As Anya delved deeper into the diary, she became consumed by its contents. She felt an inexplicable connection to Eleanor, as though their destinies were somehow intertwined.\n\nDriven by an insatiable need to know more, Anya resolved to uncover the truth behind the Hawthorne family's enigmatic past. But as she sought answers, she realized that she was playing a dangerous game. Ancient secrets have a way of coming back to life, and their consequences can be deadly..."
}

**Comment by dcpavel** [comment]
No, this change affected way more files than it should have.

File: server/types/datamodels.d.ts
Comment (dcpavel) [inline server/types/datamodels.d.ts:1]: This is a change caught by the linter.

File: server/models/NodeModel.ts
Comment (dcpavel) [inline server/models/NodeModel.ts:1]: linting change.

File: server/models/UserModel.ts
Comment (dcpavel) [inline server/models/UserModel.ts:1]: formatting change.

**Comment by dcpavel** [comment]
Reopening to attempt to remove the changes to the PULL_REQUEST_TEMPLATE.md

File: server/models/GeminiModel.ts
Comment (kinskiwu) [inline server/models/GeminiModel.ts:1]: Excellent choice in using a class to encapsulate the functionalities!

File: server/models/GeminiModel.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    getEnv();
+  }
+
+  /**
+   * Set the Gemini API key using the environment variable.
+   *
+   * @return the Gemini API key
+   */
+  setApiKey(): string {
+    try {
```
Comment (kinskiwu) [inline server/models/GeminiModel.ts:29]: Since accessing an unset environment variable with process.env.VARIABLE_NAME returns undefined instead of throwing an error, the catch block won't be executed. If you want to ensure the presence of the GEMINI_API_KEY, use a fallback and an `if` state to directly check for empty string and throw an error.

File: server/controllers/genAiController.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import createAi from '@/server/utils/aiFactory';
+import { createError } from '@/server/middleware/errorHandling';
+
+import type { Request, Response, NextFunction } from 'express';
+import type { AiInterface, AiTypes } from '@/server/types/ai';
+
+const genAiController = {
```
Comment (kinskiwu) [inline server/controllers/genAiController.ts:7]: Consider using AI instead of ai or Ai throughout the codebase to improve consistency.

File: server/types/ai.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import type { GoogleGenerativeAI, InputContent } from '@google/generative-ai';
+
+/**
+ * The interface for the AI.
+ * This will need to modified if we add more AI models.
+ */
+export interface AiInterface {
```
Comment (kinskiwu) [inline server/types/ai.d.ts:7]: The names `AiInterface` and `AiTypes` seem a bit vague. Consider using more descriptive names like AIModelService, AIProviderService, AIModel, AIProvider, or other appropriate names to clearly convey their purposes.

File: server/models/GeminiModel.ts (hunk 1 above)
Comment (dcpavel) [reply server/models/GeminiModel.ts:29]: awesome catch, thank you.

File: server/controllers/genAiController.ts (hunk 2 above)
Comment (dcpavel) [reply server/controllers/genAiController.ts:7]: I will change throughout, for consistency.

File: server/types/ai.d.ts (hunk 3 above)
Comment (dcpavel) [reply server/types/ai.d.ts:7]: I will consider alternative names (including the ones provided). I will add comments here before committing changes.

Comment (dcpavel) [reply server/types/ai.d.ts:7]: What about:
AiInterface -> GenerativeAIModel
AiTypes -> AIProvider

**Review by kinskiwu — APPROVED** [review]
Great work Darren!

File: server/controllers/genAiController.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    try {
+      const AIConnection = genAIController.createConnection(AI);
+
+      const response = await AIConnection.generateResponse(prompt);
+
+      res.locals.response = response;
+
+      return next();
+    } catch (err: unknown) {
+      console.log('here');
```
Comment (LucieSeidler) [inline server/controllers/genAiController.ts:36]: I'm a bit confused by the console logs and the github action warnings. Can we have an offline talk regarding them?

File: server/models/NodeModel.ts [hunk 5]
```diff
@@ -1,10 +1,10 @@
-import { Schema, model, Types } from "mongoose";
-import { TreeNode } from "../types/datamodels";
+import { Schema, model, Types } from 'mongoose';
```
Comment (LucieSeidler) [inline server/models/NodeModel.ts:1]: this should have been updated. Can you ensure it's not in order to minimize touching unrelated files?

File: server/types/middleware.d.ts
Comment (LucieSeidler) [inline server/types/middleware.d.ts:1]: can we have a better name for this file since I don't think this will hold all middleware definitions for Linkta

File: server/controllers/genAiController.ts (hunk 4 above)
Comment (dcpavel) [reply server/controllers/genAiController.ts:36]: This particular error (and some of the ones following them) crept in and should have been removed as I tried to resolve Linkta-196. I have removed these, but the GitHub action warnings still exist. We should have this conversation on a broader scope.

File: server/models/NodeModel.ts (hunk 5 above)
Comment (dcpavel) [reply server/models/NodeModel.ts]: This branch needs merging with the work that has been pushed to dev, I am updating it now.

File: server/types/middleware.d.ts
Comment (dcpavel) [reply server/types/middleware.d.ts:1]: I have renamed it to errorHandling.d.ts in order to better link it to the file in which it is used.

File: server/controllers/genAiController.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import createAI from '@/server/utils/AIFactory';
+import { createError } from '@/server/middleware/errorHandling';
+
+import type { Request, Response, NextFunction } from 'express';
+import { type GenerativeAIModel } from '@server/types/index';
```
Comment (zhiyi15) [inline server/controllers/genAiController.ts:5]: Not that this really matters that much, but should the 'type' be outside of the  brackets?

**Review by zhiyi15 — APPROVED** [review]
Looks great to me! Thank you Darren!!
